### ISSUE TITLE
[css-logical-1] Order of inheritance vs. mapping in logical properties

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#3029]
Example from @FremyCompany:

```
  <div style="margin-left: 10px; margin-inline-start: 20px; direction: rtl">
    <div style="margin-right: inherit; margin-inline-start: inherit; direction: ltr"></div>
  </div>
```

It's clear that `margin-right` inherits 20px from the computed `margin-right` value on the parent. What about `margin-inline-start`, though? Does it inherit 10px or 20px?

**Comment by fantasai** [comment]
Just wanted to note here that the possible solutions to #3030 depend on what happens here.

**Comment by fantasai** [comment]
> I would hope, at least, that those two examples produce the same results (respectively) as these two.

Yes, they would have to: inheritance uses the parent’s computed value, and pairs of corresponding logical/physical properties are required to have the same computed value.

**Comment by Loirooriol** [comment]
> What about `margin-inline-start`, though? Does it inherit 10px or 20px?

Chromium, WebKit and Firefox are interoperable, it inherits 10px. So if `margin-inline-start` resolves to `margin-left`, it inherits the `margin-left` of the parent, even if that's the `margin-inline-end` of the parent.

**Comment by fantasai** [comment]
@Loirooriol Thanks! We should probably check with i18n that it's acceptable behavior, but as long as it is, speccing interop sounds reasonable to me.

**Comment by r12a** [comment]
fwiw, last Friday i had created about 100 tests for css-logical, including the various directional permutations.  It's geek week now, so i may upload the tests with results this week, or it may have to wait until next week.  I might be able to add a test for this scenario, too.

**Comment by r12a** [comment]
So the main body of tests and results is at https://w3c.github.io/i18n-tests/results/css-logical

I didn't yet integrate the tests for inheritance into that page, because i'm not sure about the result.  You can find 2 tests at 
- https://w3c.github.io/i18n-tests/css-logical/inheritance/inherit-logical-logical.html
- https://w3c.github.io/i18n-tests/css-logical/inheritance/inherit-physical-logical.html

The results are indeed interoperable across Firefox, Chrome, and Safari browsers.

However, i wrote the assertions following the logic that seemed sensible to me: that is that the margin-inline-start for a LTR block (ie. on the left) would inherit the value for the margin on the _right_ (ie. at the start) of a parent RTL block.  Under that criterion, all the tests fail.

It seems somewhat illogical, as it were, that what is being inherited is not the start (ie. logical) value.

**Comment by astearns** [comment]
@r12a would you be available for any of our remaining meetings this week to discuss this issue?

https://wiki.csswg.org/planning/tpac-2020#schedule

**Comment by fantasai** [comment]
![logical-inheritance](https://user-images.githubusercontent.com/725717/96610701-c5fa6e80-12b0-11eb-950c-504ddaf93089.jpeg)

So these are the two operations under considerations:

- Blue Line: Child’s `margin-inline-start` maps to `margin-right` which then inherits from parent’s `margin-right` (which is the parent’s `margin-inline-end`). This is what browsers currently do.
- Red Line: Child’s `margin-inline-start` (which is the child’s `margin-right`) inherits from parent’s `margin-inline-start` (which is matched with parent’s `margin-left`). This is what @r12a expects.

Keep in mind that properties are first cascaded (all together, logical and physical mixed and mapped by writing mode, latest declaration of a conflicting set winning), and then computed (which causes each matching pair to share a single computed value).

**Comment by Loirooriol** [comment]
Note that if logical properties inherit from logical properties, we need to define what happens with `all: inherit`, since the order will matter. For backwards compatibility it should either not include logical properties, or it include them before physical ones.

I think it would be simpler to just keep what's currently implemented (inherit from the resolved physical properties).

**Comment by FremyCompany** [comment]
@Loirooriol We resolved yesterday that shorthand properties only apply to one type of physical-or-logical mapping.

> During parsing of a css declaration block, shorthands (like margin and all) expands only to the set of longhand (grouped by mapping logic: so either logical or physical) required to describe their value, and by default resort to use physical properties if both sets are possible. Note that they can only expand if they do not contain var() substitutions.

**Comment by Loirooriol** [comment]
https://github.com/w3c/csswg-drafts/blob/a7448f63e83e9d8f629ba66d186da26978c9664d/css-logical-1/Overview.bs#L400-L403

@fantasai I don't really get this. Which "omitted values"? "otherwise" which condition?

**Comment by fantasai** [comment]
@Loirooriol If there's a shorthand that can set both physical and logical values, e.g. suppose we extended `margin` in such a fashion, then whichever longhands are omitted (not set explicitly) should be set first so that the explicit values reset whatever is necessary. So if someone specifies `margin: relative 1em` then we set the logical longhands last so they can win against the physical ones. If we always set physical last, then setting logical values in such a shorthand would be impossible.

**Comment by Loirooriol** [comment]
@fantasai But according to #3030,

> During parsing of a css declaration block, shorthands (like margin and all) expands only to the set of longhand (grouped by mapping logic: so either logical or physical)

while you seem to mean that both sets of longhand are set? What is supposed to happen if I do

```js
document.body.style.cssText = "margin: 0";
document.body.style.length; // 4 or 8 ??
```

**Comment by FremyCompany** [comment]
My understanding is 4, the logical properties only. But this is understable since such property doesn't exist yet...

**Comment by Loirooriol** [comment]
Well, `margin` already exists. When not including the non-stable `logical` keyword, should it only expand to the physical properties only, or should it also reset the logical ones beforehand? From @fantasai's explanation the spec seems to want the latter, but I'm not sure if this is web compatible.

**Comment by Loirooriol** [comment]
5 years later, nobody has implemented [this resolution](https://github.com/w3c/csswg-drafts/issues/3029#issuecomment-715414021), and I'm not sure it would even be web compatible at this point.

Agenda+ to re-evaluate.
